// @coinbase/cdp-sdk/_esm/openapi-client/generated/x402-facilitator/x402-facilitator.js
import { cdpApiClient } from "../../cdpApiClient.js";
/**
 * Verify an x402 protocol payment with a specific scheme and network.
 * @summary Verify payment
 */
export const verifyX402Payment = (verifyX402PaymentBody, options) => {
    return cdpApiClient({
        url: `/v2/x402/verify`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: verifyX402PaymentBody,
    }, options);
};
/**
 * Settle an x402 protocol payment with a specific scheme and network.
 * @summary Settle payment
 */
export const settleX402Payment = (settleX402PaymentBody, options) => {
    return cdpApiClient({
        url: `/v2/x402/settle`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: settleX402PaymentBody,
    }, options);
};
/**
 * Get the supported x402 protocol payment schemes and networks that the facilitator is able to verify and settle payments for.
 * @summary Get supported payment schemes and networks
 */
export const supportedX402PaymentKinds = (options) => {
    return cdpApiClient({ url: `/v2/x402/supported`, method: "GET" }, options);
};
/**
 * Lists all active discovered x402 resources.
This endpoint returns resources that have been discovered and cached by the x402 facilitator, including their payment requirements and metadata.
The response is paginated, and by default, returns 100 items per page.
 * @summary List x402 resources
 */
export const listX402DiscoveryResources = (params, options) => {
    return cdpApiClient({ url: `/v2/x402/discovery/resources`, method: "GET", params }, options);
};
/**
 * Gets x402 merchant discovery information for a given merchant payment address.
This endpoint returns all active x402 resources associated with the specified `payTo` address, allowing clients to discover what payment-gated resources a merchant exposes and their corresponding payment requirements.
The response is paginated, and by default, returns 20 items per page.
 * @summary List merchant discovery info
 */
export const listX402DiscoveryMerchant = (params, options) => {
    return cdpApiClient({ url: `/v2/x402/discovery/merchant`, method: "GET", params }, options);
};
/**
 * Searches for active x402 resources using a text query and optional filters.
Supports both text-based and vector-based search depending on availability. Results are sorted by relevance and quality score.
Legacy network names (e.g., `base`, `base-sepolia`, `solana`) are automatically normalized to their CAIP-2 equivalents.
The response is limited to 20 items per request. If more results exist, `partialResults` will be `true`.
 * @summary Search x402 resources
 */
export const searchX402Resources = (params, options) => {
    return cdpApiClient({ url: `/v2/x402/discovery/search`, method: "GET", params }, options);
};
/**
 * Handles JSON-RPC requests for the Model Context Protocol (MCP). Supports MCP methods for discovering x402 payment resources and tools.
 * @summary Handle MCP JSON-RPC request
 */
export const postX402DiscoveryMcp = (x402McpRequest, options) => {
    return cdpApiClient({
        url: `/v2/x402/discovery/mcp`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: x402McpRequest,
    }, options);
};
