// @coinbase/cdp-sdk/_esm/openapi-client/generated/deposit-destinations/deposit-destinations.js
import { cdpApiClient } from "../../cdpApiClient.js";
/**
 * List deposit destinations. You can optionally filter the results by type, account ID, network, or cryptocurrency address. Results are sorted by creation date in descending order (newest first).
 * @summary List deposit destinations
 */
export const listDepositDestinations = (params, options) => {
    return cdpApiClient({ url: `/v2/deposit-destinations`, method: "GET", params }, options);
};
/**
 * Create a new deposit destination for an account. A deposit destination is a cryptocurrency address that can be used to receive funds. The address will be generated for the specified network.
 * @summary Create deposit destination
 */
export const createDepositDestination = (createDepositDestinationRequest, options) => {
    return cdpApiClient({
        url: `/v2/deposit-destinations`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: createDepositDestinationRequest,
    }, options);
};
/**
 * Get a specific deposit destination by its ID.
 * @summary Get deposit destination
 */
export const getDepositDestinationById = (depositDestinationId, options) => {
    return cdpApiClient({ url: `/v2/deposit-destinations/${depositDestinationId}`, method: "GET" }, options);
};
