// @coinbase/cdp-sdk/_esm/openapi-client/generated/end-user-accounts/end-user-accounts.js
import { cdpApiClient } from "../../cdpApiClient.js";
/**
 * Signs a transaction with the given end user EVM account.
The transaction should be serialized as a hex string using [RLP](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/).

The transaction must be an [EIP-1559 dynamic fee transaction](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md). The developer is responsible for ensuring that the unsigned transaction is valid, as the API will not validate the transaction.
 * @summary Sign transaction via end user EVM account
 */
export const signEvmTransactionWithEndUserAccount = (userId, signEvmTransactionWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/sign/transaction`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: signEvmTransactionWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs a transaction with the given end user EVM account and sends it to the indicated supported network. This API handles nonce management and gas estimation, leaving the developer to provide only the minimal set of fields necessary to send the transaction. The transaction should be serialized as a hex string using [RLP](https://ethereum.org/en/developers/docs/data-structures-and-encoding/rlp/).

The transaction must be an [EIP-1559 dynamic fee transaction](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md).


**Transaction fields and API behavior**

- `to` *(Required)*: The address of the contract or account to send the transaction to.
- `chainId` *(Ignored)*: The value of the `chainId` field in the transaction is ignored.
  The transaction will be sent to the network indicated by the `network` field in the request body.

- `nonce` *(Optional)*: The nonce to use for the transaction. If not provided, the API will assign
   a nonce to the transaction based on the current state of the account.

- `maxPriorityFeePerGas` *(Optional)*: The maximum priority fee per gas to use for the transaction.
   If not provided, the API will estimate a value based on current network conditions.

- `maxFeePerGas` *(Optional)*: The maximum fee per gas to use for the transaction.
   If not provided, the API will estimate a value based on current network conditions.

- `gasLimit` *(Optional)*: The gas limit to use for the transaction. If not provided, the API will estimate a value
  based on the `to` and `data` fields of the transaction.

- `value` *(Optional)*: The amount of ETH, in wei, to send with the transaction.
- `data` *(Optional)*: The data to send with the transaction; only used for contract calls.
- `accessList` *(Optional)*: The access list to use for the transaction.
 * @summary Send transaction via end user EVM account
 */
export const sendEvmTransactionWithEndUserAccount = (userId, sendEvmTransactionWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/send/transaction`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: sendEvmTransactionWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Sends USDC from an end user's EVM account (EOA or Smart Account) to a recipient address on a supported EVM network. This endpoint simplifies USDC transfers by automatically handling contract resolution, decimal conversion, gas estimation, and transaction encoding.
The `amount` field accepts human-readable amounts as decimal strings (e.g., "1.5", "25.50").
 * @summary Send USDC on EVM
 */
export const sendEvmAssetWithEndUserAccount = (userId, address, asset, sendEvmAssetWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/${address}/send/${asset}`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: sendEvmAssetWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs an [EIP-191](https://eips.ethereum.org/EIPS/eip-191) message with the given end user EVM account.

Per the specification, the message in the request body is prepended with `0x19 <0x45 (E)> <thereum Signed Message:\n" + len(message)>` before being signed.
 * @summary Sign EIP-191 message via end user EVM account
 */
export const signEvmMessageWithEndUserAccount = (userId, signEvmMessageWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/sign/message`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: signEvmMessageWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs [EIP-712](https://eips.ethereum.org/EIPS/eip-712) typed data with the given end user EVM account.
 * @summary Sign EIP-712 typed data via end user EVM account
 */
export const signEvmTypedDataWithEndUserAccount = (userId, signEvmTypedDataWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/sign/typed-data`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: signEvmTypedDataWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Returns the active delegation for the specified end user, if one exists. This operation can be performed by the end user themselves or by a developer using their API key.
 * @summary Get delegation for end user
 */
export const getDelegationForEndUser = (userId, params, options) => {
    return cdpApiClient({ url: `/v2/embedded-wallet-api/end-users/${userId}/delegation`, method: "GET", params }, options);
};
/**
 * Revokes all active delegations for the specified end user. This operation can be performed by the end user themselves or by a developer using their API key.
 * @summary Revoke delegation for end user
 */
export const revokeDelegationForEndUser = (userId, revokeDelegationForEndUserBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/delegation`,
        method: "DELETE",
        headers: { "Content-Type": "application/json" },
        data: revokeDelegationForEndUserBody,
        params,
    }, options);
};
/**
 * Creates an account-scoped delegation that allows a developer to sign on behalf of an end user for a single blockchain account (identified by its address) for the specified duration. The end user must be authenticated to authorize this delegation.
Multiple account-scoped delegations may exist concurrently for a single end user (one per canonical account address). Account-scoped and user-scoped delegations cannot coexist for the same user.
When the address corresponds to an EVM Smart Account, the delegation is scoped to the Smart Account's owner EOA rather than the Smart Account address itself. This means `/address/{smartAccountAddress}/delegation` and `/address/{ownerEoaAddress}/delegation` resolve to the same delegation, and the 409 `account_scoped_delegation_active` error may be returned when creating via either address if one already exists for the canonical owner.
 * @summary Create account-scoped delegation for end user
 */
export const createDelegationForEndUserAccount = (userId, address, createDelegationForEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/address/${address}/delegation`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: createDelegationForEndUserAccountBody,
        params,
    }, options);
};
/**
 * Returns the active account-scoped delegation for the specified end user account, if one exists. Useful for showing delegation status in a UI.
When the address corresponds to an EVM Smart Account, this returns the delegation for the Smart Account's owner EOA.
 * @summary Get account-scoped delegation for end user
 */
export const getDelegationForEndUserAccount = (userId, address, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/address/${address}/delegation`,
        method: "GET",
        params,
    }, options);
};
/**
 * Revokes the active account-scoped delegation for the specified end user account. Other account-scoped delegations for the same user are unaffected. This operation can be performed by the end user themselves or by a developer using their API key.
When the address corresponds to an EVM Smart Account, this revokes the delegation for the Smart Account's owner EOA.
 * @summary Revoke account-scoped delegation for end user
 */
export const revokeDelegationForEndUserAccount = (userId, address, revokeDelegationForEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/address/${address}/delegation`,
        method: "DELETE",
        headers: { "Content-Type": "application/json" },
        data: revokeDelegationForEndUserAccountBody,
        params,
    }, options);
};
/**
 * Creates an EIP-7702 delegation for an end user's EVM EOA account, upgrading it with smart account capabilities.

This endpoint:
- Retrieves delegation artifacts from onchain
- Signs the EIP-7702 authorization for delegation
- Assembles and submits a Type 4 transaction
- Creates an associated smart account object

The delegation allows the EVM EOA to be used as a smart account, which enables batched transactions and gas sponsorship via paymaster.
 * @summary Create EIP-7702 delegation for end user EVM account
 */
export const createEvmEip7702DelegationWithEndUserAccount = (userId, createEvmEip7702DelegationWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/eip7702/delegation`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: createEvmEip7702DelegationWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Prepares, signs, and sends a user operation for an end user's Smart Account.
 * @summary Send user operation for end user Smart Account
 */
export const sendUserOperationWithEndUserAccount = (userId, address, sendUserOperationWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/evm/smart-accounts/${address}/send`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: sendUserOperationWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs an arbitrary Base64 encoded message with the given Solana account.
 **WARNING:**  Never sign a message that you didn't generate as it may put your funds at risk.
 * @summary Sign Base64-encoded message
 */
export const signSolanaMessageWithEndUserAccount = (userId, signSolanaMessageWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/solana/sign/message`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: signSolanaMessageWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs a transaction with the given end user Solana account.
The unsigned transaction should be serialized into a byte array and then encoded as base64.
**Transaction types**
The following transaction types are supported:
* [Legacy transactions](https://solana-labs.github.io/solana-web3.js/classes/Transaction.html)
* [Versioned transactions](https://solana-labs.github.io/solana-web3.js/classes/VersionedTransaction.html)
The developer is responsible for ensuring that the unsigned transaction is valid, as the API will not validate the transaction.
 * @summary Sign transaction via end user Solana account
 */
export const signSolanaTransactionWithEndUserAccount = (userId, signSolanaTransactionWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/solana/sign/transaction`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: signSolanaTransactionWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Signs a transaction with the given end user Solana account and sends it to the indicated supported network.
The API handles recent blockhash management and fee estimation, leaving the developer to provide only the minimal set of fields necessary to send the transaction.
The unsigned transaction should be serialized into a byte array and then encoded as base64.
**Transaction types**
The following transaction types are supported:
* [Legacy transactions](https://solana.com/developers/guides/advanced/versions#current-transaction-versions)
* [Versioned transactions](https://solana.com/developers/guides/advanced/versions)
**Instruction Batching**
To batch multiple operations, include multiple instructions within a single transaction. All instructions within a transaction are executed atomically - if any instruction fails, the entire transaction fails and is rolled back.
**Network Support**
The following Solana networks are supported:
* `solana` - Solana Mainnet
* `solana-devnet` - Solana Devnet
The developer is responsible for ensuring that the unsigned transaction is valid, as the API will not validate the transaction.
 * @summary Send transaction via end user Solana account
 */
export const sendSolanaTransactionWithEndUserAccount = (userId, sendSolanaTransactionWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/solana/send/transaction`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: sendSolanaTransactionWithEndUserAccountBody,
        params,
    }, options);
};
/**
 * Sends USDC from an end user's Solana account to a recipient address on the Solana network. This endpoint simplifies USDC transfers by automatically handling mint resolution, Associated Token Account (ATA) creation, decimal conversion, and transaction encoding.
The `amount` field accepts human-readable amounts as decimal strings (e.g., "1.5", "25.50").
Use the optional `createRecipientAta` parameter to control whether the sender pays for creating the recipient's Associated Token Account if it doesn't exist.
 * @summary Send USDC on Solana
 */
export const sendSolanaAssetWithEndUserAccount = (userId, address, asset, sendSolanaAssetWithEndUserAccountBody, params, options) => {
    return cdpApiClient({
        url: `/v2/embedded-wallet-api/end-users/${userId}/solana/${address}/send/${asset}`,
        method: "POST",
        headers: { "Content-Type": "application/json" },
        data: sendSolanaAssetWithEndUserAccountBody,
        params,
    }, options);
};
