// @coinbase/cdp-sdk/_esm/openapi-client/generated/coinbaseDeveloperPlatformAPIs.schemas.js
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const AccountType = {
    prime: "prime",
    business: "business",
    cdp: "cdp",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ErrorType = {
    already_exists: "already_exists",
    authorization_expired: "authorization_expired",
    bad_gateway: "bad_gateway",
    capture_expired: "capture_expired",
    client_closed_request: "client_closed_request",
    customer_not_authorized: "customer_not_authorized",
    endpoint_unavailable: "endpoint_unavailable",
    faucet_limit_exceeded: "faucet_limit_exceeded",
    forbidden: "forbidden",
    idempotency_error: "idempotency_error",
    internal_server_error: "internal_server_error",
    invalid_request: "invalid_request",
    invalid_sql_query: "invalid_sql_query",
    invalid_signature: "invalid_signature",
    malformed_transaction: "malformed_transaction",
    not_found: "not_found",
    payment_method_required: "payment_method_required",
    payment_required: "payment_required",
    settlement_failed: "settlement_failed",
    rate_limit_exceeded: "rate_limit_exceeded",
    request_canceled: "request_canceled",
    service_unavailable: "service_unavailable",
    timed_out: "timed_out",
    unauthorized: "unauthorized",
    unsupported_tos_language: "unsupported_tos_language",
    policy_violation: "policy_violation",
    policy_in_use: "policy_in_use",
    account_limit_exceeded: "account_limit_exceeded",
    network_not_tradable: "network_not_tradable",
    guest_permission_denied: "guest_permission_denied",
    guest_region_forbidden: "guest_region_forbidden",
    guest_transaction_limit: "guest_transaction_limit",
    guest_transaction_count: "guest_transaction_count",
    phone_number_verification_expired: "phone_number_verification_expired",
    document_verification_failed: "document_verification_failed",
    recipient_allowlist_violation: "recipient_allowlist_violation",
    recipient_allowlist_pending: "recipient_allowlist_pending",
    refund_expired: "refund_expired",
    travel_rules_recipient_violation: "travel_rules_recipient_violation",
    source_account_invalid: "source_account_invalid",
    target_account_invalid: "target_account_invalid",
    source_account_not_found: "source_account_not_found",
    target_account_not_found: "target_account_not_found",
    source_asset_not_supported: "source_asset_not_supported",
    target_asset_not_supported: "target_asset_not_supported",
    target_email_invalid: "target_email_invalid",
    target_onchain_address_invalid: "target_onchain_address_invalid",
    transfer_amount_invalid: "transfer_amount_invalid",
    transfer_asset_not_supported: "transfer_asset_not_supported",
    transfer_quote_expired: "transfer_quote_expired",
    insufficient_balance: "insufficient_balance",
    metadata_too_many_entries: "metadata_too_many_entries",
    metadata_key_too_long: "metadata_key_too_long",
    metadata_value_too_long: "metadata_value_too_long",
    travel_rules_field_missing: "travel_rules_field_missing",
    asset_mismatch: "asset_mismatch",
    mfa_already_enrolled: "mfa_already_enrolled",
    mfa_invalid_code: "mfa_invalid_code",
    mfa_flow_expired: "mfa_flow_expired",
    mfa_required: "mfa_required",
    mfa_not_enrolled: "mfa_not_enrolled",
    order_quote_expired: "order_quote_expired",
    order_already_filled: "order_already_filled",
    order_already_canceled: "order_already_canceled",
    account_not_ready: "account_not_ready",
    insufficient_liquidity: "insufficient_liquidity",
    insufficient_allowance: "insufficient_allowance",
    transaction_simulation_failed: "transaction_simulation_failed",
    delegation_not_found: "delegation_not_found",
    delegation_expired: "delegation_expired",
    delegation_revoked: "delegation_revoked",
    delegation_not_authorized: "delegation_not_authorized",
    delegation_not_enabled: "delegation_not_enabled",
    network_mismatch: "network_mismatch",
    already_enabled: "already_enabled",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CapabilityName = {
    custodyCrypto: "custodyCrypto",
    custodyFiat: "custodyFiat",
    custodyStablecoin: "custodyStablecoin",
    tradeCrypto: "tradeCrypto",
    tradeStablecoin: "tradeStablecoin",
    transferCrypto: "transferCrypto",
    transferFiat: "transferFiat",
    transferStablecoin: "transferStablecoin",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const AssetType = {
    fiat: "fiat",
    crypto: "crypto",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const Network = {
    base: "base",
    ethereum: "ethereum",
    solana: "solana",
    aptos: "aptos",
    arbitrum: "arbitrum",
    "arbitrum-sepolia": "arbitrum-sepolia",
    optimism: "optimism",
    polygon: "polygon",
    world: "world",
    "world-sepolia": "world-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const DepositDestinationStatus = {
    active: "active",
    inactive: "inactive",
    pending: "pending",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CryptoDepositDestinationType = {
    crypto: "crypto",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CreateCryptoDepositDestinationRequestAllOfType = {
    crypto: "crypto",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CreateCryptoDepositDestinationRequestType = {
    crypto: "crypto",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TransferStatus = {
    /** Transfer was created with `execute: true`, but is momentarily being quoted before executing _or_ the transfer was created with `execute: false`. It can be executed by calling `\/v2\/transfers\/{transferId}\/execute` with `execute: true`. */
    quoted: "quoted",
    /** Transfer is executing after being quoted. No action needed - monitor progress via the transfers webhook. */
    processing: "processing",
    /** Transfer completed successfully. */
    completed: "completed",
    /** Transfer failed. See `failureReason` for details. */
    failed: "failed",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TransferFeeType = {
    BankFee: "bank",
    ConversionFee: "conversion",
    NetworkFee: "network",
    OtherFee: "other",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TravelRuleStatus = {
    /** Additional fields are required before the transfer can proceed. */
    TravelRuleStatusIncomplete: "incomplete",
    /** All requirements are satisfied and the transfer will proceed. */
    TravelRuleStatusCompleted: "completed",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TravelRuleBeneficiaryAllOfWalletType = {
    custodial: "custodial",
    self_custody: "self_custody",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const TransferRequestAmountType = {
    target: "target",
    source: "source",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const DepositTravelRuleOriginatorWalletType = {
    /** The originator\'s wallet is held by a custodial service. */
    custodial: "custodial",
    /** The originator\'s wallet is self-custodied. */
    self_custody: "self_custody",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EmailAuthenticationType = {
    email: "email",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SmsAuthenticationType = {
    sms: "sms",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const DeveloperJWTAuthenticationType = {
    jwt: "jwt",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OAuth2ProviderType = {
    google: "google",
    apple: "apple",
    x: "x",
    telegram: "telegram",
    github: "github",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SiweAuthenticationType = {
    siwe: "siwe",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmEip7702DelegationNetwork = {
    "base-sepolia": "base-sepolia",
    base: "base",
    arbitrum: "arbitrum",
    optimism: "optimism",
    polygon: "polygon",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmUserOperationNetwork = {
    "base-sepolia": "base-sepolia",
    base: "base",
    arbitrum: "arbitrum",
    optimism: "optimism",
    zora: "zora",
    polygon: "polygon",
    bnb: "bnb",
    avalanche: "avalanche",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmUserOperationStatus = {
    pending: "pending",
    signed: "signed",
    broadcast: "broadcast",
    complete: "complete",
    dropped: "dropped",
    failed: "failed",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmEip7702DelegationOperationStatus = {
    UNSPECIFIED: "UNSPECIFIED",
    PENDING: "PENDING",
    SUBMITTED: "SUBMITTED",
    COMPLETED: "COMPLETED",
    FAILED: "FAILED",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SpendPermissionNetwork = {
    base: "base",
    "base-sepolia": "base-sepolia",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
    optimism: "optimism",
    arbitrum: "arbitrum",
    avalanche: "avalanche",
    polygon: "polygon",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmSwapsNetwork = {
    base: "base",
    ethereum: "ethereum",
    arbitrum: "arbitrum",
    optimism: "optimism",
    polygon: "polygon",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ListEvmTokenBalancesNetwork = {
    base: "base",
    "base-sepolia": "base-sepolia",
    ethereum: "ethereum",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EthValueCriterionType = {
    ethValue: "ethValue",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EthValueCriterionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmAddressCriterionType = {
    evmAddress: "evmAddress",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmAddressCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const KnownAbiType = {
    erc20: "erc20",
    erc721: "erc721",
    erc1155: "erc1155",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const AbiStateMutability = {
    pure: "pure",
    view: "view",
    nonpayable: "nonpayable",
    payable: "payable",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const AbiFunctionType = {
    function: "function",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const AbiInputType = {
    constructor: "constructor",
    error: "error",
    event: "event",
    fallback: "fallback",
    receive: "receive",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmDataParameterConditionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmDataParameterConditionListOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmDataCriterionType = {
    evmData: "evmData",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const NetUSDChangeCriterionType = {
    netUSDChange: "netUSDChange",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const NetUSDChangeCriterionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTransactionRuleOperation = {
    signEvmTransaction: "signEvmTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmNetworkCriterionType = {
    evmNetwork: "evmNetwork",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmNetworkCriterionNetworksItem = {
    "base-sepolia": "base-sepolia",
    base: "base",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
    avalanche: "avalanche",
    polygon: "polygon",
    optimism: "optimism",
    arbitrum: "arbitrum",
    "arbitrum-sepolia": "arbitrum-sepolia",
    zora: "zora",
    bnb: "bnb",
    world: "world",
    "world-sepolia": "world-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmNetworkCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEvmTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEvmTransactionRuleOperation = {
    sendEvmTransaction: "sendEvmTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmMessageCriterionType = {
    evmMessage: "evmMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmMessageRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmMessageRuleOperation = {
    signEvmMessage: "signEvmMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmTypedAddressConditionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const EvmTypedNumericalConditionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTypedDataFieldCriterionType = {
    evmTypedDataField: "evmTypedDataField",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTypedDataVerifyingContractCriterionType = {
    evmTypedDataVerifyingContract: "evmTypedDataVerifyingContract",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTypedDataVerifyingContractCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTypedDataRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmTypedDataRuleOperation = {
    signEvmTypedData: "signEvmTypedData",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolAddressCriterionType = {
    solAddress: "solAddress",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolAddressCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolValueCriterionType = {
    solValue: "solValue",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolValueCriterionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SplAddressCriterionType = {
    splAddress: "splAddress",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SplAddressCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SplValueCriterionType = {
    splValue: "splValue",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SplValueCriterionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const MintAddressCriterionType = {
    mintAddress: "mintAddress",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const MintAddressCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const KnownIdlType = {
    SystemProgram: "SystemProgram",
    TokenProgram: "TokenProgram",
    AssociatedTokenProgram: "AssociatedTokenProgram",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolDataParameterConditionOperator = {
    ">": ">",
    ">=": ">=",
    "<": "<",
    "<=": "<=",
    "==": "==",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolDataParameterConditionListOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolDataCriterionType = {
    solData: "solData",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ProgramIdCriterionType = {
    programId: "programId",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ProgramIdCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignSolTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignSolTransactionRuleOperation = {
    signSolTransaction: "signSolTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolNetworkCriterionType = {
    solNetwork: "solNetwork",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolNetworkCriterionNetworksItem = {
    "solana-devnet": "solana-devnet",
    solana: "solana",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolNetworkCriterionOperator = {
    in: "in",
    not_in: "not in",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendSolTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendSolTransactionRuleOperation = {
    sendSolTransaction: "sendSolTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SolMessageCriterionType = {
    solMessage: "solMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignSolMessageRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignSolMessageRuleOperation = {
    signSolMessage: "signSolMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmHashRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEvmHashRuleOperation = {
    signEvmHash: "signEvmHash",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const PrepareUserOperationRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const PrepareUserOperationRuleOperation = {
    prepareUserOperation: "prepareUserOperation",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendUserOperationRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendUserOperationRuleOperation = {
    sendUserOperation: "sendUserOperation",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmTransactionRuleOperation = {
    signEndUserEvmTransaction: "signEndUserEvmTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserEvmTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserEvmTransactionRuleOperation = {
    sendEndUserEvmTransaction: "sendEndUserEvmTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmMessageRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmMessageRuleOperation = {
    signEndUserEvmMessage: "signEndUserEvmMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmTypedDataRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmTypedDataRuleOperation = {
    signEndUserEvmTypedData: "signEndUserEvmTypedData",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmHashRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserEvmHashRuleOperation = {
    signEndUserEvmHash: "signEndUserEvmHash",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserSolTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserSolTransactionRuleOperation = {
    signEndUserSolTransaction: "signEndUserSolTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserSolTransactionRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserSolTransactionRuleOperation = {
    sendEndUserSolTransaction: "sendEndUserSolTransaction",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserSolMessageRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SignEndUserSolMessageRuleOperation = {
    signEndUserSolMessage: "signEndUserSolMessage",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserEvmAssetRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserEvmAssetRuleOperation = {
    sendEndUserEvmAsset: "sendEndUserEvmAsset",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserSolAssetRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserSolAssetRuleOperation = {
    sendEndUserSolAsset: "sendEndUserSolAsset",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserOperationRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEndUserOperationRuleOperation = {
    sendEndUserOperation: "sendEndUserOperation",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CreateEndUserEvmSwapRuleAction = {
    reject: "reject",
    accept: "accept",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CreateEndUserEvmSwapRuleOperation = {
    createEndUserEvmSwap: "createEndUserEvmSwap",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const PolicyScope = {
    project: "project",
    account: "account",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ListSolanaTokenBalancesNetwork = {
    solana: "solana",
    "solana-devnet": "solana-devnet",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnchainDataResultSchemaColumnsItemType = {
    String: "String",
    UInt8: "UInt8",
    UInt16: "UInt16",
    UInt32: "UInt32",
    UInt64: "UInt64",
    UInt128: "UInt128",
    UInt256: "UInt256",
    Int8: "Int8",
    Int16: "Int16",
    Int32: "Int32",
    Int64: "Int64",
    Int128: "Int128",
    Int256: "Int256",
    Float32: "Float32",
    Float64: "Float64",
    Bool: "Bool",
    Date: "Date",
    DateTime: "DateTime",
    DateTime64: "DateTime64",
    UUID: "UUID",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const WebhookEventResponseStatus = {
    pending: "pending",
    processing: "processing",
    succeeded: "succeeded",
    failed: "failed",
    retrying: "retrying",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402Version = {
    NUMBER_1: 1,
    NUMBER_2: 2,
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402BatchSettlementDepositPayloadType = {
    deposit: "deposit",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402BatchSettlementVoucherPayloadType = {
    voucher: "voucher",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402BatchSettlementRefundPayloadType = {
    refund: "refund",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402BatchSettlementClaimPayloadType = {
    claim: "claim",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402BatchSettlementSettlePayloadType = {
    settle: "settle",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402V2Network = {
    "eip155:8453": "eip155:8453",
    "eip155:84532": "eip155:84532",
    "eip155:137": "eip155:137",
    "eip155:42161": "eip155:42161",
    "eip155:480": "eip155:480",
    "eip155:4801": "eip155:4801",
    "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1": "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402V2PaymentRequirementsScheme = {
    exact: "exact",
    upto: "upto",
    "batch-settlement": "batch-settlement",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402V1Network = {
    base: "base",
    "base-sepolia": "base-sepolia",
    solana: "solana",
    "solana-devnet": "solana-devnet",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402V1PaymentPayloadScheme = {
    exact: "exact",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402V1PaymentRequirementsScheme = {
    exact: "exact",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402VerifyInvalidReason = {
    insufficient_funds: "insufficient_funds",
    invalid_scheme: "invalid_scheme",
    invalid_network: "invalid_network",
    invalid_x402_version: "invalid_x402_version",
    invalid_payment_requirements: "invalid_payment_requirements",
    invalid_payload: "invalid_payload",
    invalid_exact_evm_payload_authorization_value: "invalid_exact_evm_payload_authorization_value",
    invalid_exact_evm_payload_authorization_value_too_low: "invalid_exact_evm_payload_authorization_value_too_low",
    invalid_exact_evm_payload_authorization_valid_after: "invalid_exact_evm_payload_authorization_valid_after",
    invalid_exact_evm_payload_authorization_valid_before: "invalid_exact_evm_payload_authorization_valid_before",
    invalid_exact_evm_payload_authorization_typed_data_message: "invalid_exact_evm_payload_authorization_typed_data_message",
    invalid_exact_evm_payload_authorization_from_address_kyt: "invalid_exact_evm_payload_authorization_from_address_kyt",
    invalid_exact_evm_payload_authorization_to_address_kyt: "invalid_exact_evm_payload_authorization_to_address_kyt",
    invalid_exact_evm_payload_signature: "invalid_exact_evm_payload_signature",
    invalid_exact_evm_payload_signature_address: "invalid_exact_evm_payload_signature_address",
    invalid_exact_evm_permit2_payload_allowance_required: "invalid_exact_evm_permit2_payload_allowance_required",
    invalid_exact_evm_permit2_payload_signature: "invalid_exact_evm_permit2_payload_signature",
    invalid_exact_evm_permit2_payload_deadline: "invalid_exact_evm_permit2_payload_deadline",
    invalid_exact_evm_permit2_payload_valid_after: "invalid_exact_evm_permit2_payload_valid_after",
    invalid_exact_evm_permit2_payload_spender: "invalid_exact_evm_permit2_payload_spender",
    invalid_exact_evm_permit2_payload_recipient: "invalid_exact_evm_permit2_payload_recipient",
    invalid_exact_evm_permit2_payload_amount: "invalid_exact_evm_permit2_payload_amount",
    invalid_exact_svm_payload_transaction: "invalid_exact_svm_payload_transaction",
    invalid_exact_svm_payload_transaction_amount_mismatch: "invalid_exact_svm_payload_transaction_amount_mismatch",
    invalid_exact_svm_payload_transaction_create_ata_instruction: "invalid_exact_svm_payload_transaction_create_ata_instruction",
    invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_payee: "invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_payee",
    invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_asset: "invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_asset",
    invalid_exact_svm_payload_transaction_instructions: "invalid_exact_svm_payload_transaction_instructions",
    invalid_exact_svm_payload_transaction_instructions_length: "invalid_exact_svm_payload_transaction_instructions_length",
    invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction: "invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction",
    invalid_exact_svm_payload_transaction_instructions_compute_price_instruction: "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction",
    invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high: "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high",
    invalid_exact_svm_payload_transaction_instruction_not_spl_token_transfer_checked: "invalid_exact_svm_payload_transaction_instruction_not_spl_token_transfer_checked",
    invalid_exact_svm_payload_transaction_instruction_not_token_2022_transfer_checked: "invalid_exact_svm_payload_transaction_instruction_not_token_2022_transfer_checked",
    invalid_exact_svm_payload_transaction_not_a_transfer_instruction: "invalid_exact_svm_payload_transaction_not_a_transfer_instruction",
    invalid_exact_svm_payload_transaction_cannot_derive_receiver_ata: "invalid_exact_svm_payload_transaction_cannot_derive_receiver_ata",
    invalid_exact_svm_payload_transaction_receiver_ata_not_found: "invalid_exact_svm_payload_transaction_receiver_ata_not_found",
    invalid_exact_svm_payload_transaction_sender_ata_not_found: "invalid_exact_svm_payload_transaction_sender_ata_not_found",
    invalid_exact_svm_payload_transaction_simulation_failed: "invalid_exact_svm_payload_transaction_simulation_failed",
    invalid_exact_svm_payload_transaction_transfer_to_incorrect_ata: "invalid_exact_svm_payload_transaction_transfer_to_incorrect_ata",
    invalid_exact_svm_payload_transaction_fee_payer_included_in_instruction_accounts: "invalid_exact_svm_payload_transaction_fee_payer_included_in_instruction_accounts",
    invalid_exact_svm_payload_transaction_fee_payer_transferring_funds: "invalid_exact_svm_payload_transaction_fee_payer_transferring_funds",
    invalid_batch_settlement_evm_scheme: "invalid_batch_settlement_evm_scheme",
    invalid_batch_settlement_evm_network_mismatch: "invalid_batch_settlement_evm_network_mismatch",
    invalid_batch_settlement_evm_payload_type: "invalid_batch_settlement_evm_payload_type",
    invalid_batch_settlement_evm_channel_not_found: "invalid_batch_settlement_evm_channel_not_found",
    invalid_batch_settlement_evm_deposit_simulation_failed: "invalid_batch_settlement_evm_deposit_simulation_failed",
    invalid_batch_settlement_evm_channel_id_mismatch: "invalid_batch_settlement_evm_channel_id_mismatch",
    invalid_batch_settlement_evm_channel_state_read_failed: "invalid_batch_settlement_evm_channel_state_read_failed",
    invalid_batch_settlement_evm_cumulative_below_claimed: "invalid_batch_settlement_evm_cumulative_below_claimed",
    invalid_batch_settlement_evm_cumulative_exceeds_balance: "invalid_batch_settlement_evm_cumulative_exceeds_balance",
    invalid_batch_settlement_evm_eip2612_amount_mismatch: "invalid_batch_settlement_evm_eip2612_amount_mismatch",
    invalid_batch_settlement_evm_eip2612_asset_mismatch: "invalid_batch_settlement_evm_eip2612_asset_mismatch",
    invalid_batch_settlement_evm_eip2612_deadline_expired: "invalid_batch_settlement_evm_eip2612_deadline_expired",
    invalid_batch_settlement_evm_eip2612_invalid_format: "invalid_batch_settlement_evm_eip2612_invalid_format",
    invalid_batch_settlement_evm_eip2612_invalid_signature: "invalid_batch_settlement_evm_eip2612_invalid_signature",
    invalid_batch_settlement_evm_eip2612_owner_mismatch: "invalid_batch_settlement_evm_eip2612_owner_mismatch",
    invalid_batch_settlement_evm_eip2612_spender_mismatch: "invalid_batch_settlement_evm_eip2612_spender_mismatch",
    invalid_batch_settlement_evm_erc20_approval_asset_mismatch: "invalid_batch_settlement_evm_erc20_approval_asset_mismatch",
    invalid_batch_settlement_evm_erc20_approval_from_mismatch: "invalid_batch_settlement_evm_erc20_approval_from_mismatch",
    invalid_batch_settlement_evm_erc20_approval_invalid_format: "invalid_batch_settlement_evm_erc20_approval_invalid_format",
    invalid_batch_settlement_evm_erc20_approval_unavailable: "invalid_batch_settlement_evm_erc20_approval_unavailable",
    invalid_batch_settlement_evm_erc20_approval_wrong_spender: "invalid_batch_settlement_evm_erc20_approval_wrong_spender",
    invalid_batch_settlement_evm_erc3009_authorization_required: "invalid_batch_settlement_evm_erc3009_authorization_required",
    invalid_batch_settlement_evm_insufficient_balance: "invalid_batch_settlement_evm_insufficient_balance",
    invalid_batch_settlement_evm_deposit_payload: "invalid_batch_settlement_evm_deposit_payload",
    invalid_batch_settlement_evm_receive_authorization_signature: "invalid_batch_settlement_evm_receive_authorization_signature",
    invalid_batch_settlement_evm_refund_payload: "invalid_batch_settlement_evm_refund_payload",
    invalid_batch_settlement_evm_voucher_payload: "invalid_batch_settlement_evm_voucher_payload",
    invalid_batch_settlement_evm_voucher_signature: "invalid_batch_settlement_evm_voucher_signature",
    invalid_batch_settlement_evm_missing_eip712_domain: "invalid_batch_settlement_evm_missing_eip712_domain",
    invalid_batch_settlement_evm_payload_authorization_valid_after: "invalid_batch_settlement_evm_payload_authorization_valid_after",
    invalid_batch_settlement_evm_payload_authorization_valid_before: "invalid_batch_settlement_evm_payload_authorization_valid_before",
    invalid_batch_settlement_evm_permit2_allowance_required: "invalid_batch_settlement_evm_permit2_allowance_required",
    invalid_batch_settlement_evm_permit2_amount_mismatch: "invalid_batch_settlement_evm_permit2_amount_mismatch",
    invalid_batch_settlement_evm_permit2_authorization_required: "invalid_batch_settlement_evm_permit2_authorization_required",
    invalid_batch_settlement_evm_permit2_deadline_expired: "invalid_batch_settlement_evm_permit2_deadline_expired",
    invalid_batch_settlement_evm_permit2_invalid_signature: "invalid_batch_settlement_evm_permit2_invalid_signature",
    invalid_batch_settlement_evm_permit2_invalid_spender: "invalid_batch_settlement_evm_permit2_invalid_spender",
    invalid_batch_settlement_evm_receiver_authorizer_mismatch: "invalid_batch_settlement_evm_receiver_authorizer_mismatch",
    invalid_batch_settlement_evm_receiver_mismatch: "invalid_batch_settlement_evm_receiver_mismatch",
    invalid_batch_settlement_evm_rpc_read_failed: "invalid_batch_settlement_evm_rpc_read_failed",
    invalid_batch_settlement_evm_token_mismatch: "invalid_batch_settlement_evm_token_mismatch",
    invalid_batch_settlement_evm_withdraw_delay_mismatch: "invalid_batch_settlement_evm_withdraw_delay_mismatch",
    invalid_batch_settlement_evm_withdraw_delay_out_of_range: "invalid_batch_settlement_evm_withdraw_delay_out_of_range",
    invalid_exact_evm_scheme: "invalid_exact_evm_scheme",
    invalid_exact_evm_network_mismatch: "invalid_exact_evm_network_mismatch",
    invalid_exact_evm_payload: "invalid_exact_evm_payload",
    invalid_exact_evm_payload_missing_signature: "invalid_exact_evm_payload_missing_signature",
    invalid_exact_evm_failed_to_get_network_config: "invalid_exact_evm_failed_to_get_network_config",
    invalid_exact_evm_missing_eip712_domain: "invalid_exact_evm_missing_eip712_domain",
    invalid_exact_evm_recipient_mismatch: "invalid_exact_evm_recipient_mismatch",
    invalid_exact_evm_authorization_value: "invalid_exact_evm_authorization_value",
    invalid_exact_evm_required_amount: "invalid_exact_evm_required_amount",
    invalid_exact_evm_payload_authorization_value_mismatch: "invalid_exact_evm_payload_authorization_value_mismatch",
    invalid_exact_evm_failed_to_check_nonce: "invalid_exact_evm_failed_to_check_nonce",
    invalid_exact_evm_nonce_already_used: "invalid_exact_evm_nonce_already_used",
    invalid_exact_evm_failed_to_get_balance: "invalid_exact_evm_failed_to_get_balance",
    invalid_exact_evm_insufficient_balance: "invalid_exact_evm_insufficient_balance",
    invalid_exact_evm_signature_format: "invalid_exact_evm_signature_format",
    invalid_exact_evm_failed_to_verify_signature: "invalid_exact_evm_failed_to_verify_signature",
    invalid_exact_evm_signature: "invalid_exact_evm_signature",
    invalid_exact_evm_token_name_mismatch: "invalid_exact_evm_token_name_mismatch",
    invalid_exact_evm_token_version_mismatch: "invalid_exact_evm_token_version_mismatch",
    invalid_exact_evm_eip3009_not_supported: "invalid_exact_evm_eip3009_not_supported",
    invalid_exact_evm_transaction_simulation_failed: "invalid_exact_evm_transaction_simulation_failed",
    invalid_exact_evm_verification_failed: "invalid_exact_evm_verification_failed",
    invalid_exact_evm_failed_to_parse_signature: "invalid_exact_evm_failed_to_parse_signature",
    invalid_exact_evm_failed_to_check_deployment: "invalid_exact_evm_failed_to_check_deployment",
    invalid_exact_evm_failed_to_execute_transfer: "invalid_exact_evm_failed_to_execute_transfer",
    invalid_exact_evm_failed_to_get_receipt: "invalid_exact_evm_failed_to_get_receipt",
    invalid_exact_evm_transaction_failed: "invalid_exact_evm_transaction_failed",
    invalid_exact_evm_payload_undeployed_smart_wallet: "invalid_exact_evm_payload_undeployed_smart_wallet",
    smart_wallet_deployment_failed: "smart_wallet_deployment_failed",
    unsupported_payload_type: "unsupported_payload_type",
    invalid_erc20_approval_extension_format: "invalid_erc20_approval_extension_format",
    erc20_approval_tx_failed: "erc20_approval_tx_failed",
    erc20_approval_from_mismatch: "erc20_approval_from_mismatch",
    erc20_approval_asset_mismatch: "erc20_approval_asset_mismatch",
    erc20_approval_spender_not_permit2: "erc20_approval_spender_not_permit2",
    erc20_approval_tx_parse_failed: "erc20_approval_tx_parse_failed",
    erc20_approval_tx_wrong_target: "erc20_approval_tx_wrong_target",
    erc20_approval_tx_wrong_selector: "erc20_approval_tx_wrong_selector",
    erc20_approval_tx_wrong_spender: "erc20_approval_tx_wrong_spender",
    erc20_approval_tx_signer_mismatch: "erc20_approval_tx_signer_mismatch",
    erc20_approval_tx_invalid_signature: "erc20_approval_tx_invalid_signature",
    invalid_exact_evm_unsupported_scheme: "invalid_exact_evm_unsupported_scheme",
    invalid_exact_evm_extra_field: "invalid_exact_evm_extra_field",
    invalid_exact_evm_payload_recipient_mismatch: "invalid_exact_evm_payload_recipient_mismatch",
    invalid_exact_evm_insufficient_funds: "invalid_exact_evm_insufficient_funds",
    invalid_exact_evm_transaction_state: "invalid_exact_evm_transaction_state",
    invalid_permit2_spender: "invalid_permit2_spender",
    invalid_permit2_recipient_mismatch: "invalid_permit2_recipient_mismatch",
    permit2_deadline_expired: "permit2_deadline_expired",
    permit2_not_yet_valid: "permit2_not_yet_valid",
    permit2_amount_mismatch: "permit2_amount_mismatch",
    permit2_token_mismatch: "permit2_token_mismatch",
    invalid_permit2_signature: "invalid_permit2_signature",
    permit2_allowance_required: "permit2_allowance_required",
    permit2_invalid_amount: "permit2_invalid_amount",
    permit2_invalid_destination: "permit2_invalid_destination",
    permit2_invalid_owner: "permit2_invalid_owner",
    permit2_payment_too_early: "permit2_payment_too_early",
    permit2_invalid_nonce: "permit2_invalid_nonce",
    permit2_2612_amount_mismatch: "permit2_2612_amount_mismatch",
    permit2_simulation_failed: "permit2_simulation_failed",
    permit2_insufficient_balance: "permit2_insufficient_balance",
    permit2_proxy_not_deployed: "permit2_proxy_not_deployed",
    erc20_approval_insufficient_eth_for_gas: "erc20_approval_insufficient_eth_for_gas",
    erc20_approval_broadcast_failed: "erc20_approval_broadcast_failed",
    invalid_exact_solana_unsupported_scheme: "invalid_exact_solana_unsupported_scheme",
    invalid_exact_solana_network_mismatch: "invalid_exact_solana_network_mismatch",
    invalid_exact_solana_payload_missing_fee_payer: "invalid_exact_solana_payload_missing_fee_payer",
    invalid_exact_solana_fee_payer_not_managed_by_facilitator: "invalid_exact_solana_fee_payer_not_managed_by_facilitator",
    invalid_exact_solana_payload_transaction: "invalid_exact_solana_payload_transaction",
    invalid_exact_solana_payload_transaction_could_not_be_decoded: "invalid_exact_solana_payload_transaction_could_not_be_decoded",
    invalid_exact_solana_payload_transaction_instructions_length: "invalid_exact_solana_payload_transaction_instructions_length",
    invalid_exact_solana_payload_unknown_fourth_instruction: "invalid_exact_solana_payload_unknown_fourth_instruction",
    invalid_exact_solana_payload_unknown_fifth_instruction: "invalid_exact_solana_payload_unknown_fifth_instruction",
    invalid_exact_solana_payload_unknown_sixth_instruction: "invalid_exact_solana_payload_unknown_sixth_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_limit_instruction: "invalid_exact_solana_payload_transaction_instructions_compute_limit_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_price_instruction: "invalid_exact_solana_payload_transaction_instructions_compute_price_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_price_instruction_too_high: "invalid_exact_solana_payload_transaction_instructions_compute_price_instruction_too_high",
    invalid_exact_solana_payload_no_transfer_instruction: "invalid_exact_solana_payload_no_transfer_instruction",
    invalid_exact_solana_payload_transaction_fee_payer_transferring_funds: "invalid_exact_solana_payload_transaction_fee_payer_transferring_funds",
    invalid_exact_solana_payload_mint_mismatch: "invalid_exact_solana_payload_mint_mismatch",
    invalid_exact_solana_payload_recipient_mismatch: "invalid_exact_solana_payload_recipient_mismatch",
    invalid_exact_solana_payload_amount_insufficient: "invalid_exact_solana_payload_amount_insufficient",
    invalid_exact_solana_invalid_fee_payer: "invalid_exact_solana_invalid_fee_payer",
    invalid_exact_solana_transaction_signing_failed: "invalid_exact_solana_transaction_signing_failed",
    invalid_exact_solana_transaction_simulation_failed: "invalid_exact_solana_transaction_simulation_failed",
    invalid_exact_solana_payload_memo_mismatch: "invalid_exact_solana_payload_memo_mismatch",
    invalid_exact_solana_payload_memo_count: "invalid_exact_solana_payload_memo_count",
    invalid_exact_solana_verification_failed: "invalid_exact_solana_verification_failed",
    invalid_exact_solana_fee_payer_mismatch: "invalid_exact_solana_fee_payer_mismatch",
    invalid_exact_solana_transaction_failed: "invalid_exact_solana_transaction_failed",
    invalid_exact_solana_transaction_confirmation_failed: "invalid_exact_solana_transaction_confirmation_failed",
    duplicate_settlement: "duplicate_settlement",
    invalid_exact_solana_extra_field: "invalid_exact_solana_extra_field",
    batch_settlement_cumulative_amount_mismatch: "batch_settlement_cumulative_amount_mismatch",
    batch_settlement_channel_busy: "batch_settlement_channel_busy",
    missing_batch_settlement_channel: "missing_batch_settlement_channel",
    batch_settlement_charge_exceeds_signed_cumulative: "batch_settlement_charge_exceeds_signed_cumulative",
    batch_settlement_refund_no_balance: "batch_settlement_refund_no_balance",
    batch_settlement_refund_amount_invalid: "batch_settlement_refund_amount_invalid",
    batch_settlement_refund_amount_exceeds_balance: "batch_settlement_refund_amount_exceeds_balance",
    amount_too_low: "amount_too_low",
    invalid_amount: "invalid_amount",
    kyt_risk_detected: "kyt_risk_detected",
    permit2_disabled: "permit2_disabled",
    preflight_validation_failed: "preflight_validation_failed",
    request_blocked_by_location: "request_blocked_by_location",
    self_send_not_allowed: "self_send_not_allowed",
    invalid_bazaar_extension: "invalid_bazaar_extension",
    unknown_error: "unknown_error",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402SettleErrorReason = {
    insufficient_funds: "insufficient_funds",
    invalid_scheme: "invalid_scheme",
    invalid_network: "invalid_network",
    invalid_x402_version: "invalid_x402_version",
    invalid_payment_requirements: "invalid_payment_requirements",
    invalid_payload: "invalid_payload",
    invalid_exact_evm_payload_authorization_value: "invalid_exact_evm_payload_authorization_value",
    invalid_exact_evm_payload_authorization_value_too_low: "invalid_exact_evm_payload_authorization_value_too_low",
    invalid_exact_evm_payload_authorization_valid_after: "invalid_exact_evm_payload_authorization_valid_after",
    invalid_exact_evm_payload_authorization_valid_before: "invalid_exact_evm_payload_authorization_valid_before",
    invalid_exact_evm_payload_authorization_typed_data_message: "invalid_exact_evm_payload_authorization_typed_data_message",
    invalid_exact_evm_payload_authorization_from_address_kyt: "invalid_exact_evm_payload_authorization_from_address_kyt",
    invalid_exact_evm_payload_authorization_to_address_kyt: "invalid_exact_evm_payload_authorization_to_address_kyt",
    invalid_exact_evm_payload_signature: "invalid_exact_evm_payload_signature",
    invalid_exact_evm_payload_signature_address: "invalid_exact_evm_payload_signature_address",
    invalid_exact_evm_permit2_payload_allowance_required: "invalid_exact_evm_permit2_payload_allowance_required",
    invalid_exact_evm_permit2_payload_signature: "invalid_exact_evm_permit2_payload_signature",
    invalid_exact_evm_permit2_payload_deadline: "invalid_exact_evm_permit2_payload_deadline",
    invalid_exact_evm_permit2_payload_valid_after: "invalid_exact_evm_permit2_payload_valid_after",
    invalid_exact_evm_permit2_payload_spender: "invalid_exact_evm_permit2_payload_spender",
    invalid_exact_evm_permit2_payload_recipient: "invalid_exact_evm_permit2_payload_recipient",
    invalid_exact_evm_permit2_payload_amount: "invalid_exact_evm_permit2_payload_amount",
    invalid_exact_svm_payload_transaction: "invalid_exact_svm_payload_transaction",
    invalid_exact_svm_payload_transaction_amount_mismatch: "invalid_exact_svm_payload_transaction_amount_mismatch",
    invalid_exact_svm_payload_transaction_create_ata_instruction: "invalid_exact_svm_payload_transaction_create_ata_instruction",
    invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_payee: "invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_payee",
    invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_asset: "invalid_exact_svm_payload_transaction_create_ata_instruction_incorrect_asset",
    invalid_exact_svm_payload_transaction_instructions: "invalid_exact_svm_payload_transaction_instructions",
    invalid_exact_svm_payload_transaction_instructions_length: "invalid_exact_svm_payload_transaction_instructions_length",
    invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction: "invalid_exact_svm_payload_transaction_instructions_compute_limit_instruction",
    invalid_exact_svm_payload_transaction_instructions_compute_price_instruction: "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction",
    invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high: "invalid_exact_svm_payload_transaction_instructions_compute_price_instruction_too_high",
    invalid_exact_svm_payload_transaction_instruction_not_spl_token_transfer_checked: "invalid_exact_svm_payload_transaction_instruction_not_spl_token_transfer_checked",
    invalid_exact_svm_payload_transaction_instruction_not_token_2022_transfer_checked: "invalid_exact_svm_payload_transaction_instruction_not_token_2022_transfer_checked",
    invalid_exact_svm_payload_transaction_not_a_transfer_instruction: "invalid_exact_svm_payload_transaction_not_a_transfer_instruction",
    invalid_exact_svm_payload_transaction_cannot_derive_receiver_ata: "invalid_exact_svm_payload_transaction_cannot_derive_receiver_ata",
    invalid_exact_svm_payload_transaction_receiver_ata_not_found: "invalid_exact_svm_payload_transaction_receiver_ata_not_found",
    invalid_exact_svm_payload_transaction_sender_ata_not_found: "invalid_exact_svm_payload_transaction_sender_ata_not_found",
    invalid_exact_svm_payload_transaction_simulation_failed: "invalid_exact_svm_payload_transaction_simulation_failed",
    invalid_exact_svm_payload_transaction_transfer_to_incorrect_ata: "invalid_exact_svm_payload_transaction_transfer_to_incorrect_ata",
    invalid_exact_svm_payload_transaction_fee_payer_included_in_instruction_accounts: "invalid_exact_svm_payload_transaction_fee_payer_included_in_instruction_accounts",
    invalid_exact_svm_payload_transaction_fee_payer_transferring_funds: "invalid_exact_svm_payload_transaction_fee_payer_transferring_funds",
    settle_exact_evm_transaction_confirmation_timed_out: "settle_exact_evm_transaction_confirmation_timed_out",
    settle_exact_node_failure: "settle_exact_node_failure",
    settle_exact_failed_onchain: "settle_exact_failed_onchain",
    settle_exact_svm_block_height_exceeded: "settle_exact_svm_block_height_exceeded",
    settle_exact_svm_transaction_confirmation_timed_out: "settle_exact_svm_transaction_confirmation_timed_out",
    invalid_batch_settlement_evm_unknown_settle_action: "invalid_batch_settlement_evm_unknown_settle_action",
    invalid_batch_settlement_evm_claim_payload: "invalid_batch_settlement_evm_claim_payload",
    invalid_batch_settlement_evm_settle_payload: "invalid_batch_settlement_evm_settle_payload",
    invalid_batch_settlement_evm_authorizer_address_mismatch: "invalid_batch_settlement_evm_authorizer_address_mismatch",
    invalid_batch_settlement_evm_claim_simulation_failed: "invalid_batch_settlement_evm_claim_simulation_failed",
    invalid_batch_settlement_evm_claim_transaction_failed: "invalid_batch_settlement_evm_claim_transaction_failed",
    invalid_batch_settlement_evm_deposit_transaction_failed: "invalid_batch_settlement_evm_deposit_transaction_failed",
    invalid_batch_settlement_evm_refund_simulation_failed: "invalid_batch_settlement_evm_refund_simulation_failed",
    invalid_batch_settlement_evm_refund_transaction_failed: "invalid_batch_settlement_evm_refund_transaction_failed",
    invalid_batch_settlement_evm_settle_simulation_failed: "invalid_batch_settlement_evm_settle_simulation_failed",
    invalid_batch_settlement_evm_settle_transaction_failed: "invalid_batch_settlement_evm_settle_transaction_failed",
    invalid_batch_settlement_evm_transaction_reverted: "invalid_batch_settlement_evm_transaction_reverted",
    invalid_batch_settlement_evm_wait_for_receipt_failed: "invalid_batch_settlement_evm_wait_for_receipt_failed",
    invalid_batch_settlement_evm_erc20_approval_broadcast_failed: "invalid_batch_settlement_evm_erc20_approval_broadcast_failed",
    invalid_batch_settlement_evm_channel_id_mismatch: "invalid_batch_settlement_evm_channel_id_mismatch",
    invalid_batch_settlement_evm_channel_state_read_failed: "invalid_batch_settlement_evm_channel_state_read_failed",
    invalid_batch_settlement_evm_cumulative_below_claimed: "invalid_batch_settlement_evm_cumulative_below_claimed",
    invalid_batch_settlement_evm_cumulative_exceeds_balance: "invalid_batch_settlement_evm_cumulative_exceeds_balance",
    invalid_batch_settlement_evm_eip2612_amount_mismatch: "invalid_batch_settlement_evm_eip2612_amount_mismatch",
    invalid_batch_settlement_evm_eip2612_asset_mismatch: "invalid_batch_settlement_evm_eip2612_asset_mismatch",
    invalid_batch_settlement_evm_eip2612_deadline_expired: "invalid_batch_settlement_evm_eip2612_deadline_expired",
    invalid_batch_settlement_evm_eip2612_invalid_format: "invalid_batch_settlement_evm_eip2612_invalid_format",
    invalid_batch_settlement_evm_eip2612_invalid_signature: "invalid_batch_settlement_evm_eip2612_invalid_signature",
    invalid_batch_settlement_evm_eip2612_owner_mismatch: "invalid_batch_settlement_evm_eip2612_owner_mismatch",
    invalid_batch_settlement_evm_eip2612_spender_mismatch: "invalid_batch_settlement_evm_eip2612_spender_mismatch",
    invalid_batch_settlement_evm_erc20_approval_asset_mismatch: "invalid_batch_settlement_evm_erc20_approval_asset_mismatch",
    invalid_batch_settlement_evm_erc20_approval_from_mismatch: "invalid_batch_settlement_evm_erc20_approval_from_mismatch",
    invalid_batch_settlement_evm_erc20_approval_invalid_format: "invalid_batch_settlement_evm_erc20_approval_invalid_format",
    invalid_batch_settlement_evm_erc20_approval_unavailable: "invalid_batch_settlement_evm_erc20_approval_unavailable",
    invalid_batch_settlement_evm_erc20_approval_wrong_spender: "invalid_batch_settlement_evm_erc20_approval_wrong_spender",
    invalid_batch_settlement_evm_erc3009_authorization_required: "invalid_batch_settlement_evm_erc3009_authorization_required",
    invalid_batch_settlement_evm_insufficient_balance: "invalid_batch_settlement_evm_insufficient_balance",
    invalid_batch_settlement_evm_deposit_payload: "invalid_batch_settlement_evm_deposit_payload",
    invalid_batch_settlement_evm_receive_authorization_signature: "invalid_batch_settlement_evm_receive_authorization_signature",
    invalid_batch_settlement_evm_refund_payload: "invalid_batch_settlement_evm_refund_payload",
    invalid_batch_settlement_evm_voucher_payload: "invalid_batch_settlement_evm_voucher_payload",
    invalid_batch_settlement_evm_voucher_signature: "invalid_batch_settlement_evm_voucher_signature",
    invalid_batch_settlement_evm_missing_eip712_domain: "invalid_batch_settlement_evm_missing_eip712_domain",
    invalid_batch_settlement_evm_payload_authorization_valid_after: "invalid_batch_settlement_evm_payload_authorization_valid_after",
    invalid_batch_settlement_evm_payload_authorization_valid_before: "invalid_batch_settlement_evm_payload_authorization_valid_before",
    invalid_batch_settlement_evm_permit2_allowance_required: "invalid_batch_settlement_evm_permit2_allowance_required",
    invalid_batch_settlement_evm_permit2_amount_mismatch: "invalid_batch_settlement_evm_permit2_amount_mismatch",
    invalid_batch_settlement_evm_permit2_authorization_required: "invalid_batch_settlement_evm_permit2_authorization_required",
    invalid_batch_settlement_evm_permit2_deadline_expired: "invalid_batch_settlement_evm_permit2_deadline_expired",
    invalid_batch_settlement_evm_permit2_invalid_signature: "invalid_batch_settlement_evm_permit2_invalid_signature",
    invalid_batch_settlement_evm_permit2_invalid_spender: "invalid_batch_settlement_evm_permit2_invalid_spender",
    invalid_batch_settlement_evm_receiver_authorizer_mismatch: "invalid_batch_settlement_evm_receiver_authorizer_mismatch",
    invalid_batch_settlement_evm_receiver_mismatch: "invalid_batch_settlement_evm_receiver_mismatch",
    invalid_batch_settlement_evm_rpc_read_failed: "invalid_batch_settlement_evm_rpc_read_failed",
    invalid_batch_settlement_evm_token_mismatch: "invalid_batch_settlement_evm_token_mismatch",
    invalid_batch_settlement_evm_withdraw_delay_mismatch: "invalid_batch_settlement_evm_withdraw_delay_mismatch",
    invalid_batch_settlement_evm_withdraw_delay_out_of_range: "invalid_batch_settlement_evm_withdraw_delay_out_of_range",
    invalid_batch_settlement_evm_scheme: "invalid_batch_settlement_evm_scheme",
    invalid_batch_settlement_evm_network_mismatch: "invalid_batch_settlement_evm_network_mismatch",
    invalid_batch_settlement_evm_payload_type: "invalid_batch_settlement_evm_payload_type",
    invalid_batch_settlement_evm_channel_not_found: "invalid_batch_settlement_evm_channel_not_found",
    invalid_batch_settlement_evm_deposit_simulation_failed: "invalid_batch_settlement_evm_deposit_simulation_failed",
    invalid_exact_evm_scheme: "invalid_exact_evm_scheme",
    invalid_exact_evm_network_mismatch: "invalid_exact_evm_network_mismatch",
    invalid_exact_evm_payload: "invalid_exact_evm_payload",
    invalid_exact_evm_payload_missing_signature: "invalid_exact_evm_payload_missing_signature",
    invalid_exact_evm_failed_to_get_network_config: "invalid_exact_evm_failed_to_get_network_config",
    invalid_exact_evm_missing_eip712_domain: "invalid_exact_evm_missing_eip712_domain",
    invalid_exact_evm_recipient_mismatch: "invalid_exact_evm_recipient_mismatch",
    invalid_exact_evm_authorization_value: "invalid_exact_evm_authorization_value",
    invalid_exact_evm_required_amount: "invalid_exact_evm_required_amount",
    invalid_exact_evm_payload_authorization_value_mismatch: "invalid_exact_evm_payload_authorization_value_mismatch",
    invalid_exact_evm_failed_to_check_nonce: "invalid_exact_evm_failed_to_check_nonce",
    invalid_exact_evm_nonce_already_used: "invalid_exact_evm_nonce_already_used",
    invalid_exact_evm_failed_to_get_balance: "invalid_exact_evm_failed_to_get_balance",
    invalid_exact_evm_insufficient_balance: "invalid_exact_evm_insufficient_balance",
    invalid_exact_evm_signature_format: "invalid_exact_evm_signature_format",
    invalid_exact_evm_failed_to_verify_signature: "invalid_exact_evm_failed_to_verify_signature",
    invalid_exact_evm_signature: "invalid_exact_evm_signature",
    invalid_exact_evm_token_name_mismatch: "invalid_exact_evm_token_name_mismatch",
    invalid_exact_evm_token_version_mismatch: "invalid_exact_evm_token_version_mismatch",
    invalid_exact_evm_eip3009_not_supported: "invalid_exact_evm_eip3009_not_supported",
    invalid_exact_evm_transaction_simulation_failed: "invalid_exact_evm_transaction_simulation_failed",
    invalid_exact_evm_verification_failed: "invalid_exact_evm_verification_failed",
    invalid_exact_evm_failed_to_parse_signature: "invalid_exact_evm_failed_to_parse_signature",
    invalid_exact_evm_failed_to_check_deployment: "invalid_exact_evm_failed_to_check_deployment",
    invalid_exact_evm_failed_to_execute_transfer: "invalid_exact_evm_failed_to_execute_transfer",
    invalid_exact_evm_failed_to_get_receipt: "invalid_exact_evm_failed_to_get_receipt",
    invalid_exact_evm_transaction_failed: "invalid_exact_evm_transaction_failed",
    invalid_exact_evm_payload_undeployed_smart_wallet: "invalid_exact_evm_payload_undeployed_smart_wallet",
    smart_wallet_deployment_failed: "smart_wallet_deployment_failed",
    unsupported_payload_type: "unsupported_payload_type",
    invalid_erc20_approval_extension_format: "invalid_erc20_approval_extension_format",
    erc20_approval_tx_failed: "erc20_approval_tx_failed",
    erc20_approval_from_mismatch: "erc20_approval_from_mismatch",
    erc20_approval_asset_mismatch: "erc20_approval_asset_mismatch",
    erc20_approval_spender_not_permit2: "erc20_approval_spender_not_permit2",
    erc20_approval_tx_parse_failed: "erc20_approval_tx_parse_failed",
    erc20_approval_tx_wrong_target: "erc20_approval_tx_wrong_target",
    erc20_approval_tx_wrong_selector: "erc20_approval_tx_wrong_selector",
    erc20_approval_tx_wrong_spender: "erc20_approval_tx_wrong_spender",
    erc20_approval_tx_signer_mismatch: "erc20_approval_tx_signer_mismatch",
    erc20_approval_tx_invalid_signature: "erc20_approval_tx_invalid_signature",
    invalid_exact_evm_unsupported_scheme: "invalid_exact_evm_unsupported_scheme",
    invalid_exact_evm_extra_field: "invalid_exact_evm_extra_field",
    invalid_exact_evm_payload_recipient_mismatch: "invalid_exact_evm_payload_recipient_mismatch",
    invalid_exact_evm_insufficient_funds: "invalid_exact_evm_insufficient_funds",
    invalid_exact_evm_transaction_state: "invalid_exact_evm_transaction_state",
    invalid_permit2_spender: "invalid_permit2_spender",
    invalid_permit2_recipient_mismatch: "invalid_permit2_recipient_mismatch",
    permit2_deadline_expired: "permit2_deadline_expired",
    permit2_not_yet_valid: "permit2_not_yet_valid",
    permit2_amount_mismatch: "permit2_amount_mismatch",
    permit2_token_mismatch: "permit2_token_mismatch",
    invalid_permit2_signature: "invalid_permit2_signature",
    permit2_allowance_required: "permit2_allowance_required",
    permit2_invalid_amount: "permit2_invalid_amount",
    permit2_invalid_destination: "permit2_invalid_destination",
    permit2_invalid_owner: "permit2_invalid_owner",
    permit2_payment_too_early: "permit2_payment_too_early",
    permit2_invalid_nonce: "permit2_invalid_nonce",
    permit2_2612_amount_mismatch: "permit2_2612_amount_mismatch",
    permit2_simulation_failed: "permit2_simulation_failed",
    permit2_insufficient_balance: "permit2_insufficient_balance",
    permit2_proxy_not_deployed: "permit2_proxy_not_deployed",
    erc20_approval_insufficient_eth_for_gas: "erc20_approval_insufficient_eth_for_gas",
    erc20_approval_broadcast_failed: "erc20_approval_broadcast_failed",
    invalid_exact_solana_unsupported_scheme: "invalid_exact_solana_unsupported_scheme",
    invalid_exact_solana_network_mismatch: "invalid_exact_solana_network_mismatch",
    invalid_exact_solana_payload_missing_fee_payer: "invalid_exact_solana_payload_missing_fee_payer",
    invalid_exact_solana_fee_payer_not_managed_by_facilitator: "invalid_exact_solana_fee_payer_not_managed_by_facilitator",
    invalid_exact_solana_payload_transaction: "invalid_exact_solana_payload_transaction",
    invalid_exact_solana_payload_transaction_could_not_be_decoded: "invalid_exact_solana_payload_transaction_could_not_be_decoded",
    invalid_exact_solana_payload_transaction_instructions_length: "invalid_exact_solana_payload_transaction_instructions_length",
    invalid_exact_solana_payload_unknown_fourth_instruction: "invalid_exact_solana_payload_unknown_fourth_instruction",
    invalid_exact_solana_payload_unknown_fifth_instruction: "invalid_exact_solana_payload_unknown_fifth_instruction",
    invalid_exact_solana_payload_unknown_sixth_instruction: "invalid_exact_solana_payload_unknown_sixth_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_limit_instruction: "invalid_exact_solana_payload_transaction_instructions_compute_limit_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_price_instruction: "invalid_exact_solana_payload_transaction_instructions_compute_price_instruction",
    invalid_exact_solana_payload_transaction_instructions_compute_price_instruction_too_high: "invalid_exact_solana_payload_transaction_instructions_compute_price_instruction_too_high",
    invalid_exact_solana_payload_no_transfer_instruction: "invalid_exact_solana_payload_no_transfer_instruction",
    invalid_exact_solana_payload_transaction_fee_payer_transferring_funds: "invalid_exact_solana_payload_transaction_fee_payer_transferring_funds",
    invalid_exact_solana_payload_mint_mismatch: "invalid_exact_solana_payload_mint_mismatch",
    invalid_exact_solana_payload_recipient_mismatch: "invalid_exact_solana_payload_recipient_mismatch",
    invalid_exact_solana_payload_amount_insufficient: "invalid_exact_solana_payload_amount_insufficient",
    invalid_exact_solana_invalid_fee_payer: "invalid_exact_solana_invalid_fee_payer",
    invalid_exact_solana_transaction_signing_failed: "invalid_exact_solana_transaction_signing_failed",
    invalid_exact_solana_transaction_simulation_failed: "invalid_exact_solana_transaction_simulation_failed",
    invalid_exact_solana_payload_memo_mismatch: "invalid_exact_solana_payload_memo_mismatch",
    invalid_exact_solana_payload_memo_count: "invalid_exact_solana_payload_memo_count",
    invalid_exact_solana_verification_failed: "invalid_exact_solana_verification_failed",
    invalid_exact_solana_fee_payer_mismatch: "invalid_exact_solana_fee_payer_mismatch",
    invalid_exact_solana_transaction_failed: "invalid_exact_solana_transaction_failed",
    invalid_exact_solana_transaction_confirmation_failed: "invalid_exact_solana_transaction_confirmation_failed",
    duplicate_settlement: "duplicate_settlement",
    invalid_exact_solana_extra_field: "invalid_exact_solana_extra_field",
    batch_settlement_cumulative_amount_mismatch: "batch_settlement_cumulative_amount_mismatch",
    batch_settlement_channel_busy: "batch_settlement_channel_busy",
    missing_batch_settlement_channel: "missing_batch_settlement_channel",
    batch_settlement_charge_exceeds_signed_cumulative: "batch_settlement_charge_exceeds_signed_cumulative",
    batch_settlement_refund_no_balance: "batch_settlement_refund_no_balance",
    batch_settlement_refund_amount_invalid: "batch_settlement_refund_amount_invalid",
    batch_settlement_refund_amount_exceeds_balance: "batch_settlement_refund_amount_exceeds_balance",
    amount_too_low: "amount_too_low",
    invalid_amount: "invalid_amount",
    kyt_risk_detected: "kyt_risk_detected",
    permit2_disabled: "permit2_disabled",
    preflight_validation_failed: "preflight_validation_failed",
    request_blocked_by_location: "request_blocked_by_location",
    self_send_not_allowed: "self_send_not_allowed",
    invalid_bazaar_extension: "invalid_bazaar_extension",
    unknown_error: "unknown_error",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402SupportedPaymentKindScheme = {
    exact: "exact",
    upto: "upto",
    "batch-settlement": "batch-settlement",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402SupportedPaymentKindNetwork = {
    "base-sepolia": "base-sepolia",
    base: "base",
    "solana-devnet": "solana-devnet",
    solana: "solana",
    "eip155:8453": "eip155:8453",
    "eip155:84532": "eip155:84532",
    "eip155:137": "eip155:137",
    "eip155:42161": "eip155:42161",
    "eip155:480": "eip155:480",
    "eip155:4801": "eip155:4801",
    "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp": "solana:5eykt4UsFv8P8NJdTREpY1vzqKqZKvdp",
    "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1": "solana:EtWTRABZaYq6iMfeYKouRu166VU2xqa1",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402DiscoveryResourceType = {
    http: "http",
    mcp: "mcp",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402SearchResourcesResponseSearchMethod = {
    text: "text",
    vector: "vector",
    hybrid: "hybrid",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402McpRequestJsonrpc = {
    "20": "2.0",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const X402McpResponseJsonrpc = {
    "20": "2.0",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampOrderPaymentMethodTypeId = {
    GUEST_CHECKOUT_APPLE_PAY: "GUEST_CHECKOUT_APPLE_PAY",
    GUEST_CHECKOUT_GOOGLE_PAY: "GUEST_CHECKOUT_GOOGLE_PAY",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampOrderFeeType = {
    FEE_TYPE_NETWORK: "FEE_TYPE_NETWORK",
    FEE_TYPE_EXCHANGE: "FEE_TYPE_EXCHANGE",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampOrderStatus = {
    ONRAMP_ORDER_STATUS_PENDING_AUTH: "ONRAMP_ORDER_STATUS_PENDING_AUTH",
    ONRAMP_ORDER_STATUS_PENDING_PAYMENT: "ONRAMP_ORDER_STATUS_PENDING_PAYMENT",
    ONRAMP_ORDER_STATUS_PROCESSING: "ONRAMP_ORDER_STATUS_PROCESSING",
    ONRAMP_ORDER_STATUS_COMPLETED: "ONRAMP_ORDER_STATUS_COMPLETED",
    ONRAMP_ORDER_STATUS_FAILED: "ONRAMP_ORDER_STATUS_FAILED",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampPaymentLinkType = {
    PAYMENT_LINK_TYPE_APPLE_PAY_BUTTON: "PAYMENT_LINK_TYPE_APPLE_PAY_BUTTON",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampQuotePaymentMethodTypeId = {
    CARD: "CARD",
    ACH: "ACH",
    APPLE_PAY: "APPLE_PAY",
    PAYPAL: "PAYPAL",
    FIAT_WALLET: "FIAT_WALLET",
    CRYPTO_WALLET: "CRYPTO_WALLET",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampUserIdType = {
    phone_number: "phone_number",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const OnrampLimitType = {
    weekly_spending: "weekly_spending",
    lifetime_transactions: "lifetime_transactions",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const FedwirePaymentMethodAllOfPaymentRail = {
    fedwire: "fedwire",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const FedwirePaymentMethodPaymentRail = {
    fedwire: "fedwire",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SwiftPaymentMethodAllOfPaymentRail = {
    swift: "swift",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SwiftPaymentMethodPaymentRail = {
    swift: "swift",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SepaPaymentMethodAllOfPaymentRail = {
    sepa: "sepa",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SepaPaymentMethodPaymentRail = {
    sepa: "sepa",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ListEndUsersSortItem = {
    "createdAt=asc": "createdAt=asc",
    "createdAt=desc": "createdAt=desc",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ImportEndUserBodyKeyType = {
    evm: "evm",
    solana: "solana",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEvmTransactionWithEndUserAccountBodyNetwork = {
    base: "base",
    "base-sepolia": "base-sepolia",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
    avalanche: "avalanche",
    polygon: "polygon",
    optimism: "optimism",
    arbitrum: "arbitrum",
    "arbitrum-sepolia": "arbitrum-sepolia",
    world: "world",
    "world-sepolia": "world-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEvmAssetWithEndUserAccountBodyNetwork = {
    base: "base",
    "base-sepolia": "base-sepolia",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
    avalanche: "avalanche",
    polygon: "polygon",
    optimism: "optimism",
    arbitrum: "arbitrum",
    "arbitrum-sepolia": "arbitrum-sepolia",
    world: "world",
    "world-sepolia": "world-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendSolanaTransactionWithEndUserAccountBodyNetwork = {
    solana: "solana",
    "solana-devnet": "solana-devnet",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendSolanaAssetWithEndUserAccountBodyNetwork = {
    solana: "solana",
    "solana-devnet": "solana-devnet",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendEvmTransactionBodyNetwork = {
    base: "base",
    "base-sepolia": "base-sepolia",
    ethereum: "ethereum",
    "ethereum-sepolia": "ethereum-sepolia",
    avalanche: "avalanche",
    polygon: "polygon",
    optimism: "optimism",
    arbitrum: "arbitrum",
    "arbitrum-sepolia": "arbitrum-sepolia",
    world: "world",
    "world-sepolia": "world-sepolia",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const RequestEvmFaucetBodyNetwork = {
    "base-sepolia": "base-sepolia",
    "ethereum-sepolia": "ethereum-sepolia",
    "ethereum-hoodi": "ethereum-hoodi",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const RequestEvmFaucetBodyToken = {
    eth: "eth",
    usdc: "usdc",
    eurc: "eurc",
    cbbtc: "cbbtc",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ListPoliciesScope = {
    project: "project",
    account: "account",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const CreatePolicyBodyScope = {
    project: "project",
    account: "account",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const SendSolanaTransactionBodyNetwork = {
    solana: "solana",
    "solana-devnet": "solana-devnet",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const RequestSolanaFaucetBodyToken = {
    sol: "sol",
    usdc: "usdc",
    cbtusd: "cbtusd",
};
// eslint-disable-next-line @typescript-eslint/no-redeclare
export const GetSQLSchemaDatabase = {
    base: "base",
    base_sepolia: "base_sepolia",
};
