// @rainbow-me/rainbowkit/dist/components/chunk-43UHM7WW.js
"use client";
import {
  useAsyncImage
} from "./chunk-RFBRD22G.js";
import {
  Box
} from "./chunk-ZKEPQLOV.js";

// src/components/QRCode/QRCode.tsx
import { Cuer } from "cuer";
import React from "react";

// src/components/ConnectOptions/DesktopOptions.css.ts
var QRCodeBackgroundClassName = "_1vwt0cg0";
var ScrollClassName = "_1vwt0cg2 ju367v7a ju367v7v";
var sidebar = "_1vwt0cg3";
var sidebarCompactMode = "_1vwt0cg4";

// src/components/QRCode/QRCode.tsx
function QRCode({
  ecc = "medium",
  logoBackground,
  logoSize: _logoSize = 50,
  logoUrl,
  size: sizeProp = 200,
  uri
}) {
  const padding = "20";
  const size = sizeProp - Number.parseInt(padding, 10) * 2;
  const resolvedLogoUrl = useAsyncImage(logoUrl);
  return /* @__PURE__ */ React.createElement(
    Box,
    {
      borderColor: "generalBorder",
      borderRadius: "menuButton",
      borderStyle: "solid",
      borderWidth: "1",
      className: QRCodeBackgroundClassName,
      padding,
      width: "max"
    },
    /* @__PURE__ */ React.createElement(
      Box,
      {
        style: {
          height: size,
          userSelect: "none",
          width: size
        },
        userSelect: "none"
      },
      /* @__PURE__ */ React.createElement(Cuer.Root, { errorCorrection: ecc, size, value: uri }, /* @__PURE__ */ React.createElement(
        Cuer.Cells,
        {
          className: void 0,
          fill: "currentColor",
          filter: void 0,
          radius: 1
        }
      ), /* @__PURE__ */ React.createElement(
        Cuer.Finder,
        {
          className: void 0,
          fill: "currentColor",
          radius: 0.25,
          stroke: void 0
        }
      ), resolvedLogoUrl && /* @__PURE__ */ React.createElement(Cuer.Arena, null, /* @__PURE__ */ React.createElement(
        "img",
        {
          alt: "Wallet Logo",
          src: resolvedLogoUrl,
          style: {
            backgroundColor: logoBackground,
            borderRadius: "22.5%",
            height: "88%",
            objectFit: "cover",
            width: "88%"
          }
        }
      )))
    )
  );
}

export {
  ScrollClassName,
  sidebar,
  sidebarCompactMode,
  QRCode
};
