// @rainbow-me/rainbowkit/dist/wallets/walletConnectors/chunk-SI3KAFFC.js
"use client";

// src/wallets/walletConnectors/base/base.ts
import { createConnector } from "wagmi";
import {
  baseAccount as baseAccountConnector
} from "wagmi/connectors";
var base = ({ appName, appIcon }) => {
  const { preference, ...optionalConfig } = base;
  return {
    id: "base",
    aliases: ["baseAccount"],
    name: "Base",
    shortName: "Base",
    rdns: "app.base.account",
    iconUrl: async () => (await import("./base-QS6CYWIN.js")).default,
    iconAccent: "#0000FF",
    iconBackground: "#0000FF",
    // a popup will appear prompting the user to connect or create a wallet via passkey.
    installed: true,
    createConnector: (walletDetails) => {
      const connector = baseAccountConnector({
        appName,
        appLogoUrl: appIcon,
        ...optionalConfig,
        preference: {
          telemetry: false,
          ...preference || {}
        }
      });
      return createConnector((config) => ({
        ...connector(config),
        ...walletDetails
      }));
    }
  };
};
var baseAccount = base;

export {
  base,
  baseAccount
};
