// @rainbow-me/rainbowkit/dist/wallets/walletConnectors/chunk-5WJ3FER5.js
"use client";
import {
  getWalletConnectConnector
} from "./chunk-ZRMFOB3B.js";

// src/wallets/walletConnectors/mecoWallet/mecoWallet.ts
var mecoWallet = ({
  projectId,
  walletConnectParameters
}) => ({
  id: "meco",
  name: "MeCo Wallet",
  rdns: "com.meco.wallet",
  iconUrl: async () => (await import("./mecoWallet-G3CAH6FC.js")).default,
  iconBackground: "#7443DF",
  downloadUrls: {
    android: "https://play.google.com/store/apps/details?id=com.memecore.wallet",
    ios: "https://apps.apple.com/us/app/meco-wallet/id6749523218",
    mobile: "https://mecowallet.com",
    chrome: "https://mecowallet.com",
    qrCode: "https://mecowallet.com"
  },
  mobile: {
    getUri: (uri) => {
      return `mecowallet://wc?uri=${encodeURIComponent(uri)}`;
    }
  },
  qrCode: {
    getUri: (uri) => uri,
    instructions: {
      learnMoreUrl: "https://mecowallet.com/",
      steps: [
        {
          description: "wallet_connectors.meco.qr_code.step1.description",
          step: "install",
          title: "wallet_connectors.meco.qr_code.step1.title"
        },
        {
          description: "wallet_connectors.meco.qr_code.step2.description",
          step: "create",
          title: "wallet_connectors.meco.qr_code.step2.title"
        },
        {
          description: "wallet_connectors.meco.qr_code.step3.description",
          step: "scan",
          title: "wallet_connectors.meco.qr_code.step3.title"
        }
      ]
    }
  },
  createConnector: getWalletConnectConnector({
    projectId,
    walletConnectParameters
  })
});

export {
  mecoWallet
};
