// @rainbow-me/rainbowkit/dist/wallets/walletConnectors/chunk-YRAQ7HPA.js
"use client";
import {
  getInjectedConnector,
  hasInjectedProvider
} from "./chunk-O2RJXDTM.js";

// src/wallets/walletConnectors/anchorageDigitalWallet/anchorageDigitalWallet.ts
var anchorageDigitalWallet = () => ({
  id: "anchorageDigital",
  name: "Anchorage Digital",
  rdns: "com.anchorage.connect",
  iconUrl: async () => (await import("./anchorageDigitalWallet-IWHXJ7QN.js")).default,
  iconAccent: "#FFFFFF",
  iconBackground: "#000000",
  installed: hasInjectedProvider({ flag: "isAnchorageDigital" }),
  downloadUrls: {
    chrome: "https://chromewebstore.google.com/detail/anchorage-digital/fadcgekaahkbmhjnpdhldednpbfpmkfl",
    ios: "https://apps.apple.com/us/app/anchorage-digital/id6754656781",
    mobile: "https://apps.apple.com/us/app/anchorage-digital/id6754656781",
    qrCode: "https://apps.apple.com/us/app/anchorage-digital/id6754656781"
  },
  extension: {
    instructions: {
      learnMoreUrl: "https://www.anchorage.com",
      steps: [
        {
          description: "wallet_connectors.anchorage_digital.extension.step1.description",
          step: "install",
          title: "wallet_connectors.anchorage_digital.extension.step1.title"
        },
        {
          description: "wallet_connectors.anchorage_digital.extension.step2.description",
          step: "scan",
          title: "wallet_connectors.anchorage_digital.extension.step2.title"
        },
        {
          description: "wallet_connectors.anchorage_digital.extension.step3.description",
          step: "refresh",
          title: "wallet_connectors.anchorage_digital.extension.step3.title"
        }
      ]
    }
  },
  createConnector: getInjectedConnector({ flag: "isAnchorageDigital" })
});

export {
  anchorageDigitalWallet
};
